hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/CompactingMemStore.java

```diff
@@ -0,0 +1,565 @@
+    // set sequence ids by decsending order
+    Iterator<SegmentScanner> iterator = list.descendingIterator();
+    int seqId = 0;
+    while (iterator.hasNext()) {
+      iterator.next().setSequenceID(seqId);
+      seqId++;
+    }
+    return list;
+  }
+
+  /**
+   * Check whether anything need to be done based on the current active set size.
+   * The method is invoked on every addition to the active set.
+   * For CompactingMemStore, flush the active set to the read-only memory if it's
+   * size is above threshold
+   */
+  @Override
+  protected void checkActiveSize() {
+    if (shouldFlushInMemory()) {
+      /* The thread is dispatched to flush-in-memory. This cannot be done
+      * on the same thread, because for flush-in-memory we require updatesLock
+      * in exclusive mode while this method (checkActiveSize) is invoked holding updatesLock
+      * in the shared mode. */
+      ExecutorService pool = getPool();
+      if(pool != null) {
+        InMemoryFlushWorker worker = new InMemoryFlushWorker();
+        LOG.info("Dispatching the MemStore in-memory flush for store "
+            + store.getColumnFamilyName());
+        pool.submit(worker);
+        inMemoryFlushInProgress.set(true);
+      }
+    }
+  }
+
+  // internal method, visible only for tests
+  // when invoked directly from tests it must be verified that the caller doesn't hold updatesLock,
+  // otherwise there is a deadlock
+  void flushInMemory() throws IOException {
+    // Phase I: Update the pipeline
+    getRegionServices().blockUpdates();
+    try {
+      MutableSegment active = getActive();
+      LOG.info("IN-MEMORY FLUSH: Pushing active segment into compaction pipeline, " +
+          "and initiating compaction.");
+      pushActiveToPipeline(active);
+    } finally {
+      getRegionServices().unblockUpdates();
+    }
+    // Phase II: Compact the pipeline
+    try {
+      if (allowCompaction.get()) {
+        // setting the inMemoryFlushInProgress flag again for the case this method is invoked
+        // directly (only in tests) in the common path setting from true to true is idempotent
+        inMemoryFlushInProgress.set(true);
+        // Speculative compaction execution, may be interrupted if flush is forced while
+        // compaction is in progress
+        compactor.startCompact();
+      }
+    } catch (IOException e) {
+      LOG.warn("Unable to run memstore compaction. region "
+          + getRegionServices().getRegionInfo().getRegionNameAsString()
+          + "store: "+ Bytes.toString(getFamilyName()), e);
+    }
+  }
+
+  private byte[] getFamilyName() {
+    return store.getFamily().getName();
+  }
+
+  private ExecutorService getPool() {
+    RegionServerServices rs = getRegionServices().getRegionServerServices();
+    if(rs==null) return null;
+    return rs.getExecutorService();
+  }
+
+  private boolean shouldFlushInMemory() {
+    if(getActive().getSize() > 0.9*flushSizeLowerBound) { // size above flush threshold
```

Comment:
1. Please define the constant early on in the class. 
2.  Consider using a ternary operator.